src/main/java/com/mixpanel/android/viewcrawler/ViewVisitor.java

```diff
@@ -119,11 +123,35 @@ protected String name() {
         private final Object[] mOriginalValueHolder;
     }
 
+    public static class LayoutUpdateException extends IllegalStateException {
+        public LayoutUpdateException(String message, JSONObject info) {
+            super(message);
+            mErrorInfo = info;
+        }
+
+        public JSONObject getErrorInfo() {
+            return mErrorInfo;
+        }
+
+        private JSONObject mErrorInfo;
+    }
+
     public static class LayoutUpdateVisitor extends ViewVisitor {
-        public LayoutUpdateVisitor(List<Pathfinder.PathElement> path, LayoutRule args) {
+        public LayoutUpdateVisitor(List<Pathfinder.PathElement> path, LayoutRule args, String name) {
             super(path);
             mOriginalValues = new WeakHashMap<View, LayoutRule>();
             mArgs = args;
+            mName = name;
+
+            mHorizontalRules = new ArrayList<>(Arrays.asList(
```

Comment:
Can we make it static? I think if you make it static, we can access it by ViewVisitor.LayoutUpdateVisitor.mHorizontalRules and this variable might not be initialized (unless we initialize the variable in the place where it's declared.

Make it a set could be a good idea.